src/l2_tracker/events.h

```diff
@@ -72,6 +74,63 @@ struct NewServiceNode : L2StateChange {
     static constexpr std::string_view description = "new SN"sv;
 };
```

Comment:
Is to plan to get rid of the NewServiceNode and Contributor structs above this?  There is a whole lot of duplication here between the old ones and the new ones.

Thinking ahead of the implications across a hardfork, I don't like the precedent this establishes that every time we change an event we're going to need to duplicate an event struct like this because those structs can never go away, and are duplicating rather a lot.

I'm thinking instead that we could get away with just one NewServiceNode struct here that uses the version to distinguish between contributor types and auto-converts between old contributor values and new ones depending on the version:
```c++
struct Contributor_old {
    eth::address address;
    uint64_t amount;
    
    Contributor_old(const eth::address& addr, uint64_t amt) : address{addr}, amount{amt} {}
    explicit Contributor_old(const Contributor& c) : address{c.address}, amount{c.amount} {}

    template <class Archive>
    void serialize_object(Archive& ar) {
        field(ar, "address", address);
        field_varint(ar, "amount", amount);
    }

    operator Contributor() const {
        return {address, address, amount};
        //return {address, std::nullopt, amount}; // with std::optional suggestion
    }
};

struct NewServiceNode {
    enum class Version { invalid = -1, v0, v1, _count };
    Version version = Version::v1;
    // ...
    void serialize_object(Archive& ar) {
        field_varint(ar, "version", version, [](auto v) { return v > Version::invalid && v < Version:_count; });
        field_varint(ar, "chain_id", chain_id);
        field_varint(ar, "l2_height", l2_height);
        field(ar, "service_node_pubkey", sn_pubkey);
        field(ar, "bls_pubkey", bls_pubkey);
        field(ar, "signature", ed_signature);
        field_varint(ar, "fee", fee);
        if (version == Version::v0) {
            std::vector<Contributor_old> contrib;
            if (Archive::is_serializer)
                contrib = {contributors.begin(), contributors.end()};
            field("contributors", contrib);
            if (Archive::is_deserializer) {
                contributors = {contrib.begin(), contrib.end()};
        } else {
            field("contributors", contributors);
        }
    }

    // I know I've said I don't want to actually do a HF22 for this change, but if we *did* do a HF22
    // with a change like this, this is how we could deal with that conceptually without needing to
    // make a new tx type every time we change something in an event.  (There would need to
    // be some other parts that need to change block acceptance to call this as well):

    // Checks whether the deserialized object is acceptable for the given HF
    bool check_hf_validity(cryptonote::hf hf_version) const {
        if (hf_version < hf::hf21_eth_bls)
            return false;
        if (hf_version < hf::hf22_beneficiaries)
            for (auto& c : contributors)
                if (c.beneficiary != c.address)
                    return false;
        return true;
    }
};
```

Reply:
Yeah there's explosion in code with subsequent versions since there's a bunch of places you have to update. Let's avoid that and we can do that with the `version` field as you have in V2 and in the contract's log also emit a version for that event.

With a version we won't need new TXs types for modifying fields, we can just conditionally serialise based on the version on the log. It's either you have the code to serialise or or you won't in which case you need to upgrade your node (e.g. a MSNU that bumps the patch not hardfork) So to patch this up, I'll add a version to the smart contract log and the struct here.

With `check_hf_validity` we want to avoid using a HF value to decide how to serialise/validate TX extra's because every node synchronises the L2 TXs independently from each other. If a node syncs a L2 transaction into the mempool on an Oxen  HF boundary, some could serialise it as V1 if they are slow to HF and some could serialise it as V2 if they observed the HF and this would cause consensus issues between the two partitions.

All the V1s will be removed before mainnet and to relaunch stagenet without the legacy data structures. 

So I think this is sorted with a version field. To retrofit it into `NewServiceNodeV1` is a bit tedious to handle (how to figure out if the version is present and at the front of the payload or not?) and something that we're going to drop so if it's ok, I'd keep the V1s around and drop them by launch.